### PR TITLE
Initial transfert of code

### PR DESCRIPTION
**Pull request by frheault**
Transfert from personal repo (that was the first aim to the private bitbucket) of the entire BST pipeline

- Using a folder of models, perform all BS-tracking
- Performs local/pft det/prob tracking (all at once if desired)
- Segment and clean after tracking
- Deactivate/activate bundle mask, endpoints masking, orientational priors

This PR aims to just transfer the code, so if possible the testing and general ordering and naming of tasks should probably be the main focus. The documentation, help, readme, etc. should follow shortly after.

Here is some data, launching command and a singularity
_/braindata/processedData/frheault/flow_PR.zip_

### CODE REVIEW

File: USAGE [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                                    └── S2
+                                        ├── A1
+                                        │   └── *.nii.gz
+                                        └── A2
+                                            └── *.nii.gz
+
+NOTES
+
+OPTIONAL ARGUMENTS (current value)
+--atlas_anat = "PATH_TO_IMAGE"
```
Comment (GuillaumeTh) [inline USAGE]: This is not optional. It is mandatory ?

File: USAGE [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+--atlas_directory = "PATH_TO_DIRECTORY"
```
Comment (GuillaumeTh) [inline USAGE]: same

File: USAGE [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+--atlas_bundles_basename = "BUNDLE_1,BUNDLE_2"
```
Comment (GuillaumeTh) [inline USAGE]: same

File: USAGE [hunk 4]
```diff
@@ -0,0 +1,68 @@
+Pipeline for bundle-specific tractography
+================================
```
Comment (GuillaumeTh) [inline USAGE]: Add = until the end of 'tractography'

File: main.nf [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+log.info "[Recobundles options]"
+log.info "Segmentation with Recobundle: $params.recobundle"
+log.info "Whole Brain Clustering Threshold: $params.wb_clustering_thr"
+log.info "Model Clustering Threshold: $params.model_clustering_thr"
+log.info "Prunning Threshold: $params.prunning_thr"
+log.info "Outlier Removal Alpha: $params.outlier_alpha"
+log.info ""
+log.info ""
+
+if (params.root)
```
Comment (GuillaumeTh) [inline main.nf]: Why a if here ?

File: main.nf [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+{
+    log.info "Input: $params.root"
+    root = file(params.root)
+    /* Watch out, files are ordered alphabetically in channel */
+        in_data = Channel
```
Comment (GuillaumeTh) [inline main.nf]: Why all the images are not in the same channel ?

**Review by GuillaumeTh — COMMENTED** [review]
First batch of comments

File: USAGE [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+--use_bs_endpoints_mask                 Use the bundle-specific mask for termination ($use_bs_endpoints_mask)
+--bs_endpoints_mask_dilation            Dilate the endpoint mask by a factor ($bs_endpoints_mask_dilation)
+--use_tracking_mask_as_seeding          Tractography will be initialized in the tracking mask ($use_tracking_mask_as_seeding)
+
+--local_tracking                        Perform local tractography ($local_tracking)
+--pft_tracking                          Perform particle filtering tractography ($pft_tracking)
+
+--seeding                               Seeding strategy, either npv or nts ($seeding)
+--nbr_seeds                             Number of seeds to initialize, per voxel or total ($nbr_seeds)
+--algo                                  Choice between prob, det or both ($algo)
```
Comment (GuillaumeTh) [inline USAGE]: give an example if both

File: main.nf [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+                "outlier_alpha":"$params.outlier_alpha"
+                ]
+    engine = new groovy.text.SimpleTemplateEngine()
+    template = engine.createTemplate(usage.text).make(bindings)
+    print template.toString()
+    return
+}
+
+log.info "SCIL bundle specific pipeline"
+log.info "=============================="
```
Comment (GuillaumeTh) [inline main.nf]: remove one "="

File: main.nf [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+{
+    bundles_for_exclusion
+        .set{bundles_for_inclusion}
+}
+else
+{
+    bundles_for_exclusion
+        .combine(masks_for_exclusion_process, by: 0)
+        .set{bundles_masks_for_exclusion}
+    process Mask_Exclusion{
```
Comment (GuillaumeTh) [inline main.nf]: Try to use when instead of a if

File: main.nf [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+{
+    bundles_for_inclusion
+        .set{bundles_for_recobundle}
+}
+else
+{
+    bundles_for_inclusion
+        .combine(masks_for_inclusion_process, by: 0)
+        .set{bundles_masks_for_inclusion}
+    process Mask_Inclusion{
```
Comment (GuillaumeTh) [inline main.nf]: Try to use when instead of a if

File: main.nf [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+    when: 
+    params.recobundle
+    script:
+    """
+    scil_detect_streamlines_loops.py ${bundle} no_loops.trk -a 300
+    scil_outlier_rejection.py no_loops.trk  \
+        ${sid}__${bundle_name}_${algo}_${tracking_source}_cleaned.trk \
+        --alpha $params.outlier_alpha
+    """
+}
```
Comment (GuillaumeTh) [inline main.nf]: New line in the end

**Review by GuillaumeTh — CHANGES_REQUESTED** [review]
First batch of comments. Not tested yet

File: main.nf [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+    .separate(2)
+
+workflow.onComplete {
+    log.info "Pipeline completed at: $workflow.complete"
+    log.info "Execution status: ${ workflow.success ? 'OK' : 'failed' }"
+    log.info "Execution duration: $workflow.duration"
+}
+
+anat_for_registration
+    .combine(atlas_anat)
```
Comment (GuillaumeTh) [inline main.nf:123]: Please use concat.

**Review by GuillaumeTh — CHANGES_REQUESTED** [review]
Not tested yet but looks good to me ! Good job !

File: main.nf (hunk 12 above)
Comment (frheault) [reply main.nf:123]: Concat doesn't work. Did you mean to simply replace combine with concat or is there something else to do.
For context, anat_for_registration is all of my anat and atlas_anat is a single file channel

**Pull request merged**